scalopus_general/include/scalopus_general/destructor_callback.h

```diff
@@ -34,24 +34,60 @@
 
 namespace scalopus
 {
+
 /**
- * @brief Object that calls a function on destruction.
+ * @brief Templated struct to wrap a callback function to be called on the objects' destruction.
  */
-struct DestructorCallback
+template <typename Callback>
+struct DestructorCallbackImpl
 {
 public:
-  DestructorCallback(std::function<void()>&& fun) : callback_(fun)
+  DestructorCallbackImpl(Callback&& fun) : callback_(fun)
+  {
+  }
+
+  DestructorCallbackImpl(const Callback& fun) : callback_(fun)
   {
   }
 
-  ~DestructorCallback()
+  // Call it from this void function to enforce the void() signature of the callback.
+  void destroyer()
   {
-    callback_();
+      return callback_();
```

Comment:
1. Indentation
2. Why `return` if it's `void`?

Reply:
Because this enforces that the return of the callback function is a `void` type.

If the return type of `callback_()` is not `void` this line won't compile because it would return something while the return of the `destroyer` function is void.